editor.planx.uk/src/@planx/components/shared/Schema/hook.tsx

```diff
@@ -0,0 +1,49 @@
+import { FormikConfig } from "formik";
+
+import { generateInitialValues, generateValidationSchema, Schema, SchemaData, SchemaResponse } from "./model";
+
+type UseSchema = (props: {
+  schema: Schema;
+  previousValues?: SchemaResponse[];
+}) => {
+  /**
+   * Extensible Formik config which allows custom form state and submission logic
+   * @example const formik = useFormik<SchemaData & MyCustomState>(...)
+   * @example const formik = useFormik<SchemaData>(...formikConfig, onSubmit: () => {...})
+  */
+  formikConfig: Omit<FormikConfig<SchemaData>, "onSubmit">,
+  /** 
+   * A blank set of initial values matching the schema
+   * Can be if multiple responses are allowed (e.g. in the List component)
+  */
+  initialValues: SchemaResponse,
+}
+
+/**
+ * Hook which allows you to embed a group of fields, described by a schema, within another component
+ * Form state and custom logic is stored and managed within the parent component
+ */
+export const useSchema: UseSchema = ({
+  schema,
+  previousValues,
+}) => {
+  const validationSchema = generateValidationSchema(schema);
+  const initialValues = generateInitialValues(schema);
+
+  const getInitialValues = () => {
+    if (previousValues) return previousValues;
+
+    return schema.min ? [initialValues] : [];
```

Comment:
quick question: can't quite work out/remember logically why this return depends on `schema.min` ? If `initialValues` are defined, don't we want to return them regardless of the `min` set by the schema?

Reply:
Yeah this is a good question! This was a UX decision we made when working on the List component.

If `!schema.min` we don't "pre-load" an empty item into the component as it's valid to continue without adding an item. If we didn't do this check, a user would need to cancel the item and then hit continue to get past the component. Likewise, if `schema.min` is truthy, we don't want the user to have to hit "Add new item" when this is something they would always have to do.

Reply:
Thanks for reminder :bulb: